velox/functions/prestosql/ArrayFlatten.cpp

```diff
@@ -0,0 +1,238 @@
+        state.flattenArraySizes,
+        state.flattenElementsWithDictionary);
+  }
+
+  static void createFlattenElementIndices(ProcessState& state) {
+    VELOX_CHECK(!state.flattenElementsConsecutive);
+    VELOX_CHECK_NULL(state.flattenElementIndices);
+    VELOX_CHECK_NULL(state.flattenElementsWithDictionary);
+    state.flattenElementIndices =
+        allocateIndices(state.numFlattenElements, state.pool);
+    state.rawFlattenElementIndices =
+        state.flattenElementIndices->asMutable<vector_size_t>();
+    state.flattenElementIndices->setSize(
+        state.numFlattenElements * sizeof(vector_size_t));
+    state.flattenElementsWithDictionary = BaseVector::wrapInDictionary(
+        nullptr,
+        state.flattenElementIndices,
+        state.numFlattenElements,
+        state.elements);
+  }
+
+  VectorPtr flattenArray(const ProcessState& state) const {
+    return std::make_shared<velox::ArrayVector>(
+        state.pool,
+        state.resultType,
+        nullptr,
+        state.numRows,
+        state.flattenArrayOffsets,
+        state.flattenArraySizes,
+        state.elements);
+  }
+
+  void processOneInputArray(
+      vector_size_t inputRow,
+      vector_size_t outputRow,
+      ProcessState& state) const {
+    vector_size_t nextElementOffset{-1};
+    for (auto i = 0, offset = state.arrayOffsets[inputRow];
+         i < state.arraySizes[inputRow];
+         ++i, ++offset) {
+      if (FOLLY_UNLIKELY(state.decodedElementArray->isNullAt(offset))) {
+        continue;
+      }
+      const auto elementArrayIndex = state.decodedElementArray->index(offset);
+      const auto elementOffset = state.elementOffsets[elementArrayIndex];
+      const auto numElements = state.elementSizes[elementArrayIndex];
+      state.numFlattenElements += numElements;
+      state.rawFlattenArraySizes[outputRow] += numElements;
+      if (nextElementOffset == -1) {
+        state.rawFlattenArrayOffsets[outputRow] = elementOffset;
+      } else if (nextElementOffset != elementOffset) {
+        state.flattenElementsConsecutive = false;
+      }
+      nextElementOffset = elementOffset + numElements;
+    }
+  }
+
+  void processOneOutputRow(
+      vector_size_t inputRow,
+      vector_size_t outputRow,
+      const ProcessState& state,
+      vector_size_t& flattenElementOffset) const {
+    state.rawFlattenArrayOffsets[outputRow] = flattenElementOffset;
+    vector_size_t numFlattenElements{0};
+    for (auto i = 0, elementArrayOffset = state.arrayOffsets[inputRow];
+         i < state.arraySizes[inputRow];
+         ++i, ++elementArrayOffset) {
+      if (FOLLY_UNLIKELY(
+              state.decodedElementArray->isNullAt(elementArrayOffset))) {
+        continue;
+      }
+      const auto elementArrayIndex =
+          state.decodedElementArray->index(elementArrayOffset);
+      const auto elementOffset = state.elementOffsets[elementArrayIndex];
+      const auto numElements = state.elementSizes[elementArrayIndex];
```

Comment:
This code seems to be repeated twice. Wondering if it can be replaced with a method call on state.

```
const auto {elementOffset, numElements} = state.elementOffsetAndSize(elementArrayOffset);
```

This would return {0, 0} if array is null.